velox/expression/SimpleFunctionRegistry.h

```diff
@@ -55,9 +55,43 @@ struct FunctionEntry {
   const FunctionFactory factory_;
 };
 
+struct SignatureHash {
+  size_t operator()(const FunctionSignature& signature) const {
+    return std::hash<std::vector<TypeSignature>>()(signature.argumentTypes());
+  }
+};
+
+/// Only consider function signature argument types baseName.
+/// E.g. registered map<int, long> as function input argument before, now we
+/// try to register map<long, long>, we will get it from SignatureMap, but its
+/// physical type is not equal, so not overwrite map<int, long>, both map<int,
+/// long> and map<long, long> exist in FunctionEntry.
+/// Function with Variadic<int64_t> or int64_t should not exist at the same
+/// time, we cannot distinguish which to call for single input.
+struct SignatureCmp {
+  bool operator()(const FunctionSignature& l, const FunctionSignature& r)
+      const {
+    if (l.argumentTypes().size() != r.argumentTypes().size()) {
+      return false;
+    }
+    for (auto i = 0; i < l.argumentTypes().size(); ++i) {
+      // For Map(K, V), we only consider the argument type map, not consider its
+      // element type, so as Decimal, List, Struct.let the metadata
+      // physicalSignatureEquals which will consider the child data type to
+      // decide if overwrite.
+      if (l.argumentTypes()[i].baseName() != r.argumentTypes()[i].baseName()) {
+        return false;
+      }
+    }
+    return true;
+  }
+};
+
 using SignatureMap = std::unordered_map<
     FunctionSignature,
-    std::vector<std::unique_ptr<const FunctionEntry>>>;
+    std::vector<std::unique_ptr<const FunctionEntry>>,
+    SignatureHash,
```

Comment:
My understanding is:

- SignatureHash includes argument types, but not return type
- SignatureCmd includes argument types, but in a limited way (map(k, v) compares equal to map(k1, v1) no matter what k, v, k1, v1 are.

This means that 2 entries may have different hashes, but compare the same. This is a violation of a core contract for hash/compare. We cannot do that.

Let me know if I missed something.

CC: @rui-mo

Reply:
Yes, you are right.
I change the SignatureHash to only consider base name like SignatureCmd.